bip-v1-packages.mediawiki

```diff
@@ -0,0 +1,314 @@
+
+Child-with-unconfirmed-parents package relay is negotiated between two peers through the "sendpackages" message
+with version=1.
+
+[[File:./bip-v1-packages/sendpackages_negotiation.png|600px]]
+
+''Diagram: Version 1 package relay is negotiated during the version handshake.''
+
+===New Messages===
+
+A new inv type, "MSG_PCKG1", and new protocol message, "PCKGINFO1", are added.
+
+====pckginfo1====
+
+{|
+|  Field Name  ||  Type  ||  Size  ||   Purpose
+|-
+|blockhash || uint256 || 32 || The chain tip at which this package is defined.
+|-
+|pckg_fee||CAmount||4|| The sum total fees paid by all transactions in the package.
+|-
+|pckg_weight||int64_t||8|| The sum total weight of all transactions in the package.
+|-
+|txns_length||CompactSize||1 or 3 bytes|| The number of transactions provided.
+|-
+|txns||List of wtxids||txns_length * 32|| The wtxids of each transaction in the package.
+|}
+
+
+# The "pckginfo1" message has the structure defined above, with pchCommand == "pckginfo1".
+
+# A "pckginfo1" message contains information about a version 1 package (defined below), referenced by the wtxid of the transaction it pertains to and the current blockhash.
+
+# Upon receipt of a "pckginfo1" message, the node should decide if it wants to validate the package, request transaction data if necessary, etc.
+
+# Upon receipt of a malformed "pckginfo1" message or package that does not abide by the max_count, max_weight, or other rules specified by the version agreed upon in the initial negotiation, the sender should be disconnected.  If a node receives a "pckginfo1" message for which the "pckg_fee" or "pckg_weight" do not reflect the true total fees and weight, respectively, or the transactions in the package, the message is malformed.
+
+# A node MUST NOT send a "pckginfo1" message that has not been requested by the recipient. Upon receipt of an unsolicited "pckginfo1", a node should disconnect the sender.
+
+# A "pckginfo1" message should only be sent if both peers agreed to send version 1 packages in the version handshake. If a "pckginfo1" message is received from a peer with which package relay was not negotiated, the sender should be disconnected.
+
+====MSG_PCKG1====
+
+# A new inv type (MSG_PCKG1 == 0x6) is added, for use in inv messages and getdata requests pertaining to version 1 packages.
+
+# As an inv type, it indicates that both transaction data and version 1 package information are available for the transaction. The transaction is referenced by its wtxid. As a getdata request type, it indicates that the sender wants package information for the transaction.
+
+# Upon receipt of a "getdata" request for "MSG_PCKG1", the node should respond with the version 1 package corresponding to the requested transaction and its current chain tip, or with NOTFOUND. The node should not assume that the sender is requesting the transaction data as well.
```

Comment:
Is it permitted to send a GETDATA(MSG_PCKG1, wtxid) if the peer didn't send an INV(MSG_PCKG1, wtxid)?  Ie could software use this message to reconstruct the parents of an orphan transaction?

Reply:
Yes absolutely! This kind of "half solves" the orphan fetching use case if there's only direct parents missing, but I actually think it would make sense to create another type of package for tx-with-unconfirmed-ancestors:

<img width="954" alt="image" src="https://user-images.githubusercontent.com/25183001/168891185-1630f583-de47-4937-86b1-2652cf8852f2.png">

To answer the likely question of "why wasn't v1 packages tx-with-unconfirmed-ancestors?" I have two primary reasons:

- The mempool policy created for child-with-unconfirmed-parents packages is insufficient to ensure that a the fees of a tx-with-ancestors package would be evaluated correctly. I noticed you had some [consideration](https://github.com/bitcoin/bitcoin/pull/16401/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R1054-R1056) for that in #16401. Probably the best algorithm is to sort topologically, then submit each transaction with its ancestor subset within the package. But it doesn't seem like anyone is asking for more than 2 generations of fee-bumping, and I haven't really figured out what RBF would look like.

- It seems like the orphan-processing and fee-bumping use cases can be thought of separately. For instance, fee-bumping should be a sender-initiated "I'm telling you about extra info you need" and orphan-fetching is more appropriate as a receiver-initiated "hey I'm new here, so I don't know the ancestors of this transaction."